File: leetcode/algorithms/math/perfect-number.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Perfect Number"
*/

/**
 * @param {number} num
 * @return {boolean}
 */

const getFactors = (num) => {
    let i=2;
    const factorSet = new Set();
    while(i<Math.floor(num/2)){
        if(num%i===0){
            factorSet.add(i);
            factorSet.add(num/i);
        }
        ++i;
    }
    return Array.from(factorSet);
    
}
var checkPerfectNumber = function(num) {
    return getFactors(num).concat(1).filter(n => n!=num).reduce((sum,n) => sum+n,0)===num;   
};

File: leetcode/algorithms/math/perfect-number.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Perfect Number"
*/

class Solution {
    private ArrayList<Integer> getFactors(int num){
        ArrayList<Integer> factorsList = new ArrayList<Integer>(){
            {
                add(1);
            }
        };
        int i=2;
        while(i<(int)(num/2)){
            if(num%i==0){
                if(!factorsList.contains(i)){
                    factorsList.add(i);
                }
                if(!factorsList.contains((int)(num/i))){
                    factorsList.add((int)(num/i));
                }
            }
            ++i;
        }
        return factorsList;
    }
    public boolean checkPerfectNumber(int num) {
        int sum=0;
        for(Integer n : getFactors(num)){
            if(n.intValue()!=num){
                sum+=n.intValue();
            }
        }
        return sum==num;
        
    }
}